src/itemstackmetadata.cpp

```diff
@@ -1,34 +1,55 @@
 #include "itemstackmetadata.h"
 #include "util/serialize.h"
 #include "util/strfnd.h"
+#include <json/json.h>
 
-#define DESERIALIZE_START '\x01'
-#define DESERIALIZE_KV_DELIM '\x02'
-#define DESERIALIZE_PAIR_DELIM '\x03'
-#define DESERIALIZE_START_STR "\x01"
-#define DESERIALIZE_KV_DELIM_STR "\x02"
-#define DESERIALIZE_PAIR_DELIM_STR "\x03"
+#define SERIALIZATION_VERSION_IDENTIFIER '\x02'
 
 void ItemStackMetadata::serialize(std::ostream &os) const
 {
-	std::ostringstream os2;
-	os2 << DESERIALIZE_START;
+	Json::Value root(Json::objectValue);
 	for (StringMap::const_iterator
 			it = m_stringvars.begin();
 			it != m_stringvars.end(); ++it) {
-		os2 << it->first << DESERIALIZE_KV_DELIM
-		    << it->second << DESERIALIZE_PAIR_DELIM;
+		root[it->first] = it->second;
 	}
-	os << serializeJsonStringIfNeeded(os2.str());
+	os << serializeJsonStringIfNeeded(SERIALIZATION_VERSION_IDENTIFIER +
+			Json::FastWriter().write(root));
 }
 
+#define DESERIALIZE_START '\x01'
+#define DESERIALIZE_KV_DELIM '\x02'
+#define DESERIALIZE_PAIR_DELIM '\x03'
+#define DESERIALIZE_START_STR "\x01"
+#define DESERIALIZE_KV_DELIM_STR "\x02"
+#define DESERIALIZE_PAIR_DELIM_STR "\x03"
+
 void ItemStackMetadata::deSerialize(std::istream &is)
 {
 	std::string in = deSerializeJsonStringIfNeeded(is);
 
 	m_stringvars.clear();
 
-	if (!in.empty() && in[0] == DESERIALIZE_START) {
+	if (in.empty())
+		return;
+
+	if (in[0] == SERIALIZATION_VERSION_IDENTIFIER) {
+		in.erase(0, 1);
+
+		Json::Reader reader;
+		Json::Value attr_root;
+		if (reader.parse(in, attr_root) && attr_root.type() == Json::objectValue) {
+			const Json::Value::Members attr_list = attr_root.getMemberNames();
+			for (Json::Value::Members::const_iterator it = attr_list.begin();
+					it != attr_list.end(); ++it) {
+				m_stringvars[*it] = attr_root[*it].asString();
+			}
+		} else {
+			errorstream << "ItemStackMetadata::deSerialize():"
+				<< " expected JSON data, but failed to parse. Ignoring." << std::endl;
+		}
+	} else if (in[0] == DESERIALIZE_START) {
+		// BACKWARDS COMPATIBILITY
```

Comment:
lots of servers run dev builds, this stops corruption. It can be removed before 0.4.16 release